Improve numerical stability of BiCGStab on GPU

yashwantbezawada opened the pull request:
## Summary

This PR improves the numerical stability of \`jax.scipy.sparse.linalg.bicgstab\` on GPU by reordering operations in the computation of p_i and introducing an intermediate variable to make the cancellation explicit.

## Changes

Modified the BiCGStab update step from:
\`\`\`python
beta = rho_ / rho * alpha / omega
p_ = _add(r, _mul(beta, _sub(p, _mul(omega, q))))
\`\`\`
to:
\`\`\`python
gamma = rho_ / rho * alpha
p_ = _add(r, _sub(_mul(gamma / omega, p), _mul(gamma, q)))
\`\`\`

While these formulations are mathematically equivalent (since \`beta = gamma/omega\` and \`gamma = beta*omega\`), the new version provides better numerical stability on GPU by:
1. Avoiding nested operations that can compound rounding errors
2. Making the cancellation in the \`beta*omega\` term explicit through the \`gamma\` variable
3. Separating the division by \`omega\` from the multiplication

## Motivation

This change addresses issue #32978, where BiCGStab was experiencing iterative breakdown on GPU but not on CPU for certain problems. The issue was traced to floating-point precision differences between CPU and GPU, where the original nested operation order led to numerical instability.

The approach follows similar patterns used by NVIDIA's AMGX library and has been validated to resolve the breakdown issue in finite element analysis problems reported in the issue.

## Testing

- All existing BiCGStab tests pass with this change (23 passed, 8 skipped)
- The modification is mathematically equivalent, so it preserves correctness
- The reporter in #32978 has confirmed this type of reordering resolves their GPU breakdown issue

Fixes #32978

yashwantbezawada committed:
Improve numerical stability of BiCGStab on GPU

Changes the order of operations in the BiCGStab update step for p_i from:
  p_i = r + beta * (p - omega * q)
to:
  p_i = r + (gamma/omega) * p - gamma * q

where gamma = rho_i/rho_{i-1} * alpha_{i-1}.

This mathematically equivalent formulation improves numerical stability
on GPU by making the cancellation explicit: since beta = gamma/omega,
the term gamma in the second part effectively represents beta*omega,
allowing better numerical stability than the nested operations in the
original formula.

Fixes #32978

yashwantbezawada commented:
/gemini review

yashwantbezawada commented:
/gemini review

jakevdp reviewed (changes requested):
Looks reasonable. Could you add a regression test for the change? That is, a test that passes with the change you made but would fail without it?

yashwantbezawada committed:
Add regression test for BiCGStab numerical stability

Tests that BiCGStab correctly handles matrix structures that can
cause breakdown, using tight tolerances and multiple iterations to
exercise the numerical stability improvements from issue #32978.
